File: 20.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stack;
        int n = s.size();
        char c;
        for (int i = 0; i < n; i++) {
            c = s[i];
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c);
            }
            else {
                if (stack.empty()) {
                    return false;
                }
                if (c == ')') {
                    if (stack.top() != '(') {
                        return false;
                    }
                }
                if (c == '}') {
                    if (stack.top() != '{') {
                        return false;
                    }
                }
                if (c == ']') {
                    if (stack.top() != '[') {
                        return false;
                    }
                }
                stack.pop();
            }
        }
        // We shouldn't have any unclosed brackets
        return stack.empty();
    }
};

File: 20.cs
public class Solution {
    public bool IsValid(string s) {
        Stack<char> myStack = new Stack<char>();
        foreach (char c in s) {
            char d;
            switch (c) {
                case '(':
                    myStack.Push(c);
                    break;
                case ')':
                    if (myStack.Count == 0) {
                        return false;
                    }
                    d = myStack.Pop();
                    if (d != '(') {
                        return false;
                    }
                    break;
                case '[':
                    myStack.Push(c);
                    break;
                case ']':
                    if (myStack.Count == 0) {
                        return false;
                    }
                    d = myStack.Pop();
                    if (d != '[') {
                        return false;
                    }
                    break;
                case '{':
                    myStack.Push(c);
                    break;
                case '}':
                    if (myStack.Count == 0) {
                        return false;
                    }
                    d = myStack.Pop();
                    if (d != '{') {
                        return false;
                    }
                    break;
                default:
                    return false;
                    break;
            }
        }
        if (myStack.Count > 0) {
            return false;
        }
        return true;
    }
}
